### PR TITLE
[bluestore] Eliminate recursion from Onode::put()

### PR DESCRIPTION
**Pull request by aclamk**
Onode::put() removes self from onode_map.
This remove call causes us to call Onode::put() back again.
It requires the Onode::put() to employ clever strategies to handle this recursion.

Now we first extract self from onode_map, complete (logically) current put(),
and then let put() of extracted node run.

I think this change makes code more robust.

This work is in relation to 
https://tracker.ceph.com/issues/56382
but I do not claim that it solves it.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by aclamk** [comment]
@ifed01 Do you think we could benefit from this refactor?

**Comment by aclamk** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -3678,7 +3684,6 @@ void BlueStore::Onode::get() {
   }
 }
 void BlueStore::Onode::put() {
-  ++put_nref;
```
Comment (ronen-fr) [inline src/os/bluestore/BlueStore.cc:3681]: A question from the un-initiated: you are no longer guaranteeing that the nref will be positive while playing around with the entry. There is a guarantee that there are no parallel threads that might react to the nref being 0 while still manipulating the object?

**Comment by aclamk** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:3681]: The assumption is, as with any reference counting schemes, that you have to obtain ref to manipulate object.
If nref == 1 this means that the only reference is in onode_map holding OnodeRef objects.
There is only one function that can access the node now: get_onode(). It operates under the same lock that is in Onode::put().

**Comment by aclamk** [comment]
jenkins test make check

**Comment by ifed01** [comment]
> of extrac



> @ifed01 Do you think we could benefit from this refactor?

Honestly I doubt that. 
I can see two parts in your PR:
1) Postponing last onode ref decrement using _extract/hold. Which is nice but effectively just takes second put out of the lock scope. Not sure it has any benefit by itself.
2) Eliminating put_nref permits a thread to release an onode while another concurrent thread is still accessing this onode. And this looks unsafe. The idea behind put_nref is to protect onode for that interim state when onode reference is decremented but put method might still use the onode. Now it's gone and I can see no replacement for that....

**Pull request closed without merging**